File: dlna-dmc/src/main/java/com/android/cast/dlna/dmc/control/ControlImpl.java
package com.android.cast.dlna.dmc.control;

import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.android.cast.dlna.core.ICast;
import com.android.cast.dlna.core.Utils;

import org.fourthline.cling.controlpoint.ControlPoint;
import org.fourthline.cling.model.meta.Device;

import java.util.Map;

public class ControlImpl implements ICastInterface.IControl {

    private final IServiceFactory mServiceFactory;
    private final Device<?, ?, ?> mDevice;
    private final Map<String, IServiceAction.IServiceActionCallback<?>> mCallbackMap;
    private String mUri;

    public ControlImpl(@NonNull ControlPoint controlPoint, @NonNull Device<?, ?, ?> device,
                       Map<String, IServiceAction.IServiceActionCallback<?>> map, ICastInterface.ISubscriptionListener subscriptionListener) {
        mDevice = device;
        mCallbackMap = map;
        mServiceFactory = new IServiceFactory.ServiceFactoryImpl(controlPoint, device);
        ((BaseServiceExecutor) mServiceFactory.getAvService()).execute(event -> {
            if (subscriptionListener != null) {
                subscriptionListener.onSubscriptionTransportStateChanged(event);
            }
        });
        ((BaseServiceExecutor) mServiceFactory.getRenderService()).execute(event -> {
            if (subscriptionListener != null) {
                subscriptionListener.onSubscriptionTransportStateChanged(event);
            }
        });
    }

    @Override
    public void cast(Device<?, ?, ?> device, ICast object) {
        mUri = object.getUri();
        mServiceFactory.getAvService().cast(new ICastInterface.CastEventListener() {
            @Override
            public void onSuccess(String result) {
                IServiceAction.IServiceActionCallback<Object> listener = getCallback(IServiceAction.ServiceAction.CAST);
                if (listener != null) listener.onSuccess(result);
            }

            @Override
            public void onFailed(String errMsg) {
                IServiceAction.IServiceActionCallback<Object> listener = getCallback(IServiceAction.ServiceAction.CAST);
                if (listener != null) listener.onFailed(errMsg);
            }
        }, object.getUri(), Utils.getMetadata(object));
    }

    @Override
    public boolean isCasting(Device<?, ?, ?> device) {
        return mDevice.equals(device);
    }

    @Override
    public boolean isCasting(Device<?, ?, ?> device, @Nullable String uri) {
        if (TextUtils.isEmpty(uri)) return isCasting(device);
        return isCasting(device) && uri != null && uri.equals(mUri);
    }

    @Override
    public void stop() {
        mServiceFactory.getAvService().stop(getCallback(IServiceAction.ServiceAction.STOP));
    }

    @Override
    public void play() {
        mServiceFactory.getAvService().play(getCallback(IServiceAction.ServiceAction.PLAY));
    }

    @Override
    public void pause() {
        mServiceFactory.getAvService().pause(getCallback(IServiceAction.ServiceAction.PAUSE));
    }

    @Override
    public void seekTo(long position) {
        mServiceFactory.getAvService().seek(getCallback(IServiceAction.ServiceAction.SEEK_TO), position);
    }

    @Override
    public void setVolume(int percent) {
        mServiceFactory.getRenderService().setVolume(getCallback(IServiceAction.ServiceAction.SET_VOLUME), percent);
    }

    @Override
    public void setMute(boolean mute) {
        mServiceFactory.getRenderService().setMute(getCallback(IServiceAction.ServiceAction.SET_MUTE), mute);
    }

    @Override
    public void setBrightness(int percent) {
        mServiceFactory.getRenderService().setBrightness(getCallback(IServiceAction.ServiceAction.SET_BRIGHTNESS), percent);
    }

    @SuppressWarnings("unchecked")
    @Nullable
    private <T> IServiceAction.IServiceActionCallback<T> getCallback(IServiceAction.ServiceAction action) {
        Object result = mCallbackMap.get(action.name());
        if (result == null) return null;
        return (IServiceAction.IServiceActionCallback<T>) result;
    }
}

File: dlna-dmc/src/main/java/com/android/cast/dlna/dmc/control/ControlImpl.kt
package com.android.cast.dlna.dmc.control

import android.text.TextUtils
import com.android.cast.dlna.core.ICast
import com.android.cast.dlna.core.Utils.getMetadata
import com.android.cast.dlna.dmc.control.ServiceFactory.ServiceFactoryImpl
import com.android.cast.dlna.dmc.control.ServiceAction.CAST
import com.android.cast.dlna.dmc.control.ServiceAction.PAUSE
import com.android.cast.dlna.dmc.control.ServiceAction.PLAY
import com.android.cast.dlna.dmc.control.ServiceAction.SEEK_TO
import com.android.cast.dlna.dmc.control.ServiceAction.SET_BRIGHTNESS
import com.android.cast.dlna.dmc.control.ServiceAction.SET_MUTE
import com.android.cast.dlna.dmc.control.ServiceAction.SET_VOLUME
import com.android.cast.dlna.dmc.control.ServiceAction.STOP
import org.fourthline.cling.controlpoint.ControlPoint
import org.fourthline.cling.model.meta.Device
import org.fourthline.cling.support.model.TransportState

class ControlImpl(
    controlPoint: ControlPoint,
    private val device: Device<*, *, *>,
    private val callbackMap: Map<String, ServiceActionCallback<*>>,
    subscriptionListener: SubscriptionListener?,
) : Control {

    private val serviceFactory: ServiceFactory
    private var uri: String? = null

    init {
        serviceFactory = ServiceFactoryImpl(controlPoint, device)
        (serviceFactory.avService as BaseServiceExecutor).execute(object : SubscriptionListener {
            override fun onSubscriptionTransportStateChanged(event: TransportState) {
                subscriptionListener?.onSubscriptionTransportStateChanged(event)
            }
        })
        (serviceFactory.renderService as BaseServiceExecutor).execute(object : SubscriptionListener {
            override fun onSubscriptionTransportStateChanged(event: TransportState) {
                subscriptionListener?.onSubscriptionTransportStateChanged(event)
            }
        })
    }

    override fun cast(device: Device<*, *, *>, cast: ICast) {
        uri = cast.uri
        serviceFactory.avService.cast(object : CastEventListener {
            override fun onResponse(response: ActionResponse<String>) {
                getCallback<String>(CAST)?.onResponse(response)
            }
        }, cast.uri, getMetadata(cast))
    }

    override fun isCasting(device: Device<*, *, *>?): Boolean {
        return this.device == device
    }

    override fun isCasting(device: Device<*, *, *>?, uri: String?): Boolean {
        return if (TextUtils.isEmpty(uri)) isCasting(device) else isCasting(device) && uri != null && uri == this.uri
    }

    override fun stop() {
        serviceFactory.avService.stop(getCallback(STOP))
    }

    override fun play() {
        serviceFactory.avService.play(getCallback(PLAY))
    }

    override fun pause() {
        serviceFactory.avService.pause(getCallback(PAUSE))
    }

    override fun seekTo(millSeconds: Long) {
        serviceFactory.avService.seek(getCallback(SEEK_TO), millSeconds)
    }

    override fun setVolume(percent: Int) {
        serviceFactory.renderService.setVolume(getCallback(SET_VOLUME), percent)
    }

    override fun setMute(mute: Boolean) {
        serviceFactory.renderService.setMute(getCallback(SET_MUTE), mute)
    }

    override fun setBrightness(percent: Int) {
        serviceFactory.renderService.setBrightness(getCallback(SET_BRIGHTNESS), percent)
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> getCallback(action: ServiceAction): ServiceActionCallback<T>? {
        return callbackMap[action.name] as? ServiceActionCallback<T>
    }
}